### ISSUE TITLE
Stripplot color coding bug in create.boxplot

### ISSUE THREAD

**Issue by alkaZeltser** [uclahs-cds/public-R-BoutrosLab-plotting-general#4]
I seem to have stumbled upon a bug involving multiple panels of boxplots in create.boxplot().

I'm using the formula `score ~ sex | gene` to plot multiple pairs of boxplots (a pair per gene), each pair categorized by sex (male vs female). I'm using the option `add.stripplot = TRUE` to show individual data points from each distribution. I'm using a vector of color names matched to the sex of each data point to color the points accordingly `points.col = sex.colors` (male samples colored blue, female samples colored pink).

The issue: sometimes (not always) the color-coding turns out incorrectly for specific boxplot pairs (male samples colored pink, female samples colored blue). There may be correct and incorrect panels within the same plot. Please see the example plot below.
@zhuchcn can comment futher.

A sample of the dataset:
sex | cancer_code | gene | expression_score
-- | -- | -- | --
Male | LAML | DDX3X | -0.985645936
Female | LAML | EIF1AX | -2.351457551
Male | LAML | MAGEH1 | -0.194331577
Female | LAML | RHD | 0.266507057
Male | LAML | ZNF317 | -0.470812179

`sex.colors = 'blue', 'pink',  'blue', 'pink',  'blue'`

Example of the code:

    create.boxplot( 
        formula = expression_score ~ sex | gene,  
        data = expression.data,  
        add.stripplot = TRUE,  
        points.col = sex.colors,  
        layout = c(length(unique(expression.data$gene)),1)
        );
    
Full code can be viewed by lab members [here.](https://github.com/uclahs-cds/project-CancerBiology-SexDifferencesv3PedProtFunc/blob/51a821f8aab52b909682662cdf877d74f6e9572a/src/lib/functional-genomics/plotting_functions.R#L270)

Color coding bug in last two panels (RHD and ZNF13).

<img width="360" alt="Weird_stripplot_LAML" src="https://user-images.githubusercontent.com/30507223/129140004-d6c2921a-0e87-4e8e-963b-bf243955b9fe.png">

**Comment by zhuchcn** [comment]
The reason for this is here:

https://github.com/uclahs-cds/public-R-BoutrosLab-plotting-general/blob/6e861c2c61d4fa72f62a493f9aa236f65c65e2b2/R/create.boxplot.R#L206

When calling `lattice::bwplot` with `panel.stripplot`, it works well with out the group (the argument after the `|` in formula). But when there is group in formula, for each panel (group), the `x` and `y` values that `panel.stripplot` receives are subsetted from the input data, but col still uses the entire things. So the result is, for each panel, only the first `n` values of the `points.col` is used and `n` equals to the number of observations in this group.

A work around for now is to create some dummy observations in the input data with values being NA, to make sure each group has the equal number of observations with each colour. But in long term, we probably should fix this. Just paste here for a record, that the usage of `lattice::bwplot` below is able to recognize the colors correctly.

```
bwplot(x = expression_score ~ sex | gene, expression.data, groups = sex.colours, panel = "panel.stripplot")
```

**Comment by zhuchcn** [comment]
@tyamaguchi-ucla @jmlivingstone @bugh1

**Comment by alkaZeltser** [comment]
A few updates:
For future reference, the bug can be recreated entirely with the following simplified data:
[create.boxplot.data.txt](https://github.com/uclahs-cds/public-R-BoutrosLab-plotting-general/files/7003378/create.boxplot.data.txt)

And the following code:

    data = read.table(file = 'create.boxplot.data.txt', header = TRUE);
    
    # Assemble color vector
    sex.colours <- replace(data$sex, which(data$sex == 'Male'),'dodgerblue');
    sex.colours <- replace(sex.colours, which(data$sex == 'Female'), 'pink');
    # Plot
    create.boxplot(
        formula = expression_score ~ sex | gene,
        data = data,
        add.stripplot = TRUE,
        points.col = sex.colours
        );

Producing this plot:
<img width="380" alt="Screen Shot 2021-08-17 at 3 40 20 PM" src="https://user-images.githubusercontent.com/30507223/129810156-4ecb772e-5ceb-4098-a7ff-c66d985cdffb.png">

The main problem seems to be, as @zhuchcn described, that the color coding for the first group/gene is being applied to both. In the case of an ugly work-around, adding dummy observations to the data could work, but it's not enough to simply match the number of samples in each group. The order in which the samples are read in the second group is not the same as the order of the first `n` colors. So for this strategy to work, the samples from gene1 and the samples in gene2 must be in the exact same color order. Which therefore means that each gene group must be modified to have the same number of samples of each sex.

The following example data, in this order, with two added dummy observations, would probably work:
gene  | sex | score
-- | -- | --
Gene1  | Male    |  1
Gene1 |  Male   |   2
Gene1 |  Female | 3
Gene1  | Female  | NA (dummy)
Gene2  | Male   |   1
Gene2  | Male    |  NA  (dummy)
Gene2  | Female| 2
Gene2  | Female | 3

The color order has to be the same for both gene groups. In this case (blue, blue, pink, pink) would work for both.

**Comment by zhuchcn** [comment]
You just need to sort the data before plotting. See below:

```R
data <- read.table(file = 'create.boxplot.data.txt', header = TRUE);
data <- data[order(data$gene, data$sex),];
# Assemble color vector
sex.colours <- replace(data$sex, which(data$sex == 'Male'),'dodgerblue');
sex.colours <- replace(sex.colours, which(data$sex == 'Female'), 'pink');
# Plot
create.boxplot(
    formula = expression_score ~ sex | gene,
    data = data,
    add.stripplot = TRUE,
    points.col = sex.colours
    );
```

![image](https://user-images.githubusercontent.com/29579625/129812406-a90f417c-4a7d-4ce0-a96b-41d2f30d3b1e.png)

**Comment by alkaZeltser** [comment]
Not quite, I see an offending pink dot on the EIF1AX on the Male side. I think this is because the two groups don't have the same number of samples per sex.

**Comment by zhuchcn** [comment]
We just need to make sure the number of points in each block is the same across all panels. See below.

```R
data <- read.table(file = 'create.boxplot.data.txt', header = TRUE);
dummy.data <- data.frame(
    expression_score = NA,
    sex = 'Male',
    gene = 'DDX3X'
    );
data <- rbind(data, dummy.data);
data <- data[order(data$gene, data$sex),]
print(data)
# Assemble color vector
sex.colours <- replace(data$sex, which(data$sex == 'Male'),'dodgerblue');
sex.colours <- replace(sex.colours, which(data$sex == 'Female'), 'pink');
# Plot
create.boxplot(
    formula = expression_score ~ sex | gene,
    data = data,
    add.stripplot = TRUE,
    points.col = sex.colours
    );
```

Here we have 4 points of Female and 7 points of Male in both genes, and they are sorted accordinly.

|   | expression_score|sex    |gene   |
|:--|----------------:|:------|:------|
|2  |       -1.3602409|Female |DDX3X  |
|6  |       -1.4956393|Female |DDX3X  |
|7  |       -1.4200057|Female |DDX3X  |
|9  |       -1.5611668|Female |DDX3X  |
|1  |       -0.3617344|Male   |DDX3X  |
|3  |       -0.9856459|Male   |DDX3X  |
|4  |       -0.8121548|Male   |DDX3X  |
|5  |       -0.1454028|Male   |DDX3X  |
|8  |       -1.2969527|Male   |DDX3X  |
|10 |       -1.6550786|Male   |DDX3X  |
|22 |               NA|Male   |DDX3X  |
|12 |       -1.6414475|Female |EIF1AX |
|13 |       -1.8162853|Female |EIF1AX |
|15 |       -1.6609925|Female |EIF1AX |
|17 |       -1.3229956|Female |EIF1AX |
|11 |       -0.5620692|Male   |EIF1AX |
|14 |       -1.1627007|Male   |EIF1AX |
|16 |       -2.0966206|Male   |EIF1AX |
|18 |       -2.0444492|Male   |EIF1AX |
|19 |       -0.9651082|Male   |EIF1AX |
|20 |       -1.8400535|Male   |EIF1AX |
|21 |       -0.3354791|Male   |EIF1AX |

![image](https://user-images.githubusercontent.com/29579625/129813550-978b3284-22e7-440f-8a26-46a247d6ef49.png)
